### DISCUSSION TITLE
too many root nodes to train

### DISCUSSION THREAD

**Discussion by D19CrMei** [talmolab/sleap#820] [Help!]
In the bottom-up method, if multiple edges of the skeleton are not connected, will it affect the training? My GUI reminds me that there are too many root nodes to train. Is that right? Is there any way to modify it

**Comment by roomrys** [comment]
Hi @D19CrMei,

For the bottom-up model, the [skeleton](https://sleap.ai/guides/skeletons.html#:~:text=In%20bottom%2Dup%20models%2C%20the%20edges%20are%20important%20and%20must%20connect%20all%20nodes%20to%20one%20another.) needs to be a tree graph (each node has only one source and the graph is connected)
> In bottom-up models, the edges are important and must connect all nodes to one another.

Thanks,
Liezl

**Reply by D19CrMei** [reply]
"For the bottom-up model, the skeleton needs to be a tree graph", does this algorithm require a theoretical basis?
Then, in top-down, do you still need to generate PAF like bottom-up, using flow shift and Hungarian algorithm？ thanks！

**Reply by roomrys** [reply]
Hi @D19CrMei,

Bottom-up models require a tree graph for the skeleton because we use Part Affinity Fields (PAFs) to determine connections between nodes. Top-down models do not use PAFs because the grouping problem (grouping of nodes into a single skeleton) is taken care of by the neural network which can reason about where nodes are located from the centered crop about the instance.

Thanks,
Liezl

**Reply by D19CrMei** [reply]
In the SLEAP paper, the bottom-up model process is very clear, but the top-down model is very vague for the two-stage network inside, and there is no relevant model diagram and calculation formula. Can you elaborate? thanks
